tests/lib/controllers/aligned-controller.test.js

```diff
@@ -0,0 +1,936 @@
+    expect(alignment.createAlignedTexts).not.toHaveBeenCalled()
+    expect(result).toBeFalsy()
+  })
+
+  it('3 AlignedController - createAlignedTexts defines alignment and executes tokenizer', () => {
+    const alignedC = new AlignedController()
+
+    const alignment = new Alignment({ text: 'origin', direction: 'ltr', lang: 'eng' }, { text: 'target', direction: 'ltr', lang: 'eng' })
+    jest.spyOn(alignment, 'createAlignedTexts')
+    
+    const result = alignedC.createAlignedTexts(alignment)
+
+    expect(alignment.createAlignedTexts).toHaveBeenCalled()
+    expect(result).toBeTruthy()
+  })
+
+  it('6 AlignedController - originAlignedText returns empty object if alignment is not defined', () => {
+    const alignedC = new AlignedController()
+
+    const result = alignedC.originAlignedText
+    expect(result).toEqual({})
+  })
+
+  it('4 AlignedController - originAlignedText returns originAlignedText from the defined alignment', () => {
+    const alignedC = new AlignedController()
+
+    const alignment = new Alignment({ text: 'origin', direction: 'ltr', lang: 'eng' }, { text: 'target', direction: 'ltr', lang: 'eng' })
+    alignedC.createAlignedTexts(alignment)
+
+    const result = alignedC.originAlignedText
+    expect(result).toHaveProperty('textType', 'origin')
+    expect(result).toHaveProperty('direction', 'ltr')
+    expect(result).toHaveProperty('lang', 'eng')
+
+    expect(result.tokens.length).toEqual(1)
+  })
+
+  it('5 AlignedController - targetAlignedText returns empty object if alignment is not defined', () => {
+    const alignedC = new AlignedController()
+    
+    const result = alignedC.targetAlignedText
+    expect(result).toEqual({})
+  })
+
+
+  it('6 AlignedController - targetAlignedText returns targetAlignedText from the defined alignment', () => {
+    const alignedC = new AlignedController()
+
+    const alignment = new Alignment({ text: 'origin', direction: 'ltr', lang: 'eng' }, { text: 'target', direction: 'ltr', lang: 'eng' })
+    alignedC.createAlignedTexts(alignment)
+
+    const result = alignedC.targetAlignedText
+    expect(result).toHaveProperty('textType', 'target')
+    expect(result).toHaveProperty('direction', 'ltr')
+    expect(result).toHaveProperty('lang', 'eng')
+
+    expect(result.tokens.length).toEqual(1)
+  })
+
+  it('7 AlignedController - clickToken executes activateGroupByToken if there is no active alignment and token is already grouped', () => {
+    const alignedC = new AlignedController()
+    const alignment = new Alignment({ text: 'origin', direction: 'ltr', lang: 'eng' }, { text: 'target', direction: 'ltr', lang: 'eng' })
+    alignedC.createAlignedTexts(alignment)
+    
+    alignedC.activateGroupByToken = jest.fn()
+    alignedC.startNewAlignmentGroup = jest.fn()
+    alignedC.finishActiveAlignmentGroup = jest.fn()
+    alignedC.mergeActiveGroupWithAnotherByToken = jest.fn()
+    alignedC.addToAlignmentGroup = jest.fn()
+    alignment.removeFromAlignmentGroup = jest.fn()

```

Comment:
It feels a little odd to be testing the implementation of the business logic this way.  I wonder if what we should be testing here is the end result of the action, rather tightly coupling the test to how the business logic is implemented. This makes the test pretty brittle. That is, I think the test should verify that the result of clicking on the token is that we now have an active alignment of the group the token belongs to.  

This applies to many of the controller tests below.

What do you think?

Reply:
I am not agree with it - I believe (especially) in such difficult methods we should check both:
* data was changed correctly
* there were no incorrect methods that could changed the data

that's why I think that tracking this is important (especially on the development stage)

Reply:
ok, we can agree to disagree.